pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test: TestSQLStatsCompactor failed

pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test.TestSQLStatsCompactor [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9729603?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9729603?buildTab=artifacts#/) on release-23.1 @ [9724980a4395e2b4d4cc51e561a4b768f84c992b](https://github.com/cockroachdb/cockroach/commits/9724980a4395e2b4d4cc51e561a4b768f84c992b):


```
=== RUN   TestSQLStatsCompactor
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/9a30d5271a65be6b3f2aa1433be9011f/logTestSQLStatsCompactor843096810
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestSQLStatsCompactor
    compaction_test.go:262: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9a30d5271a65be6b3f2aa1433be9011f/logTestSQLStatsCompactor843096810
--- FAIL: TestSQLStatsCompactor (153.51s)
=== RUN   TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=205/rowsDeletePerTxn=0
    compaction_test.go:229: 
        	Error Trace:	pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go:229
        	Error:      	Not equal: 
        	            	expected: 24
        	            	actual  : 16
        	Test:       	TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=205/rowsDeletePerTxn=0
        	Messages:   	expected 24 number of wide scans issued, but 16 number of wide scan issued
    --- FAIL: TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=205/rowsDeletePerTxn=0 (16.75s)
=== RUN   TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1
    compaction_test.go:229: 
        	Error Trace:	pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go:229
        	Error:      	Not equal: 
        	            	expected: 32
        	            	actual  : 25
        	Test:       	TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1
        	Messages:   	expected 32 number of wide scans issued, but 25 number of wide scan issued
    --- FAIL: TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1 (39.47s)
=== RUN   TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1024
    compaction_test.go:229: 
        	Error Trace:	pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go:229
        	Error:      	Not equal: 
        	            	expected: 32
        	            	actual  : 24
        	Test:       	TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1024
        	Messages:   	expected 32 number of wide scans issued, but 24 number of wide scan issued
    --- FAIL: TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1024 (22.31s)
=== RUN   TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=2
    compaction_test.go:229: 
        	Error Trace:	pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go:229
        	Error:      	Not equal: 
        	            	expected: 32
        	            	actual  : 24
        	Test:       	TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=2
        	Messages:   	expected 32 number of wide scans issued, but 24 number of wide scan issued
    --- FAIL: TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=2 (38.09s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101767 pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test: TestSQLStatsCompactor failed [C-test-failure O-robot T-cluster-observability branch-master]
- #101674 pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test: TestSQLStatsCompactor failed [C-test-failure O-robot T-cluster-observability branch-release-23.1.0]
</p>
</details>
/cc @cockroachdb/cluster-observability
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSQLStatsCompactor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-27224

pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test.TestSQLStatsCompactor [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9745704?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9745704?buildTab=artifacts#/) on release-23.1 @ [01f129a59ab66166d6ef7189a8930f2b37cced74](https://github.com/cockroachdb/cockroach/commits/01f129a59ab66166d6ef7189a8930f2b37cced74):


```
=== RUN   TestSQLStatsCompactor
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/9a30d5271a65be6b3f2aa1433be9011f/logTestSQLStatsCompactor4016202747
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestSQLStatsCompactor
    compaction_test.go:262: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9a30d5271a65be6b3f2aa1433be9011f/logTestSQLStatsCompactor4016202747
--- FAIL: TestSQLStatsCompactor (298.19s)
=== RUN   TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1
    compaction_test.go:229: 
        	Error Trace:	pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go:229
        	Error:      	Not equal: 
        	            	expected: 32
        	            	actual  : 25
        	Test:       	TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1
        	Messages:   	expected 32 number of wide scans issued, but 25 number of wide scan issued
    --- FAIL: TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1 (114.06s)
=== RUN   TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=2
    compaction_test.go:229: 
        	Error Trace:	pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go:229
        	Error:      	Not equal: 
        	            	expected: 32
        	            	actual  : 29
        	Test:       	TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=2
        	Messages:   	expected 32 number of wide scans issued, but 29 number of wide scan issued
    --- FAIL: TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=2 (94.38s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101767 pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test: TestSQLStatsCompactor failed [C-test-failure O-robot T-cluster-observability branch-master]
- #101674 pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test: TestSQLStatsCompactor failed [C-test-failure O-robot T-cluster-observability branch-release-23.1.0]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSQLStatsCompactor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test.TestSQLStatsCompactor [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9811164?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9811164?buildTab=artifacts#/) on release-23.1 @ [25b70262ef10a01a17aedc7b8014ad7d44de0bc9](https://github.com/cockroachdb/cockroach/commits/25b70262ef10a01a17aedc7b8014ad7d44de0bc9):


```
=== RUN   TestSQLStatsCompactor
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/9a30d5271a65be6b3f2aa1433be9011f/logTestSQLStatsCompactor561227524
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestSQLStatsCompactor
    compaction_test.go:262: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9a30d5271a65be6b3f2aa1433be9011f/logTestSQLStatsCompactor561227524
--- FAIL: TestSQLStatsCompactor (135.00s)
=== RUN   TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=205/rowsDeletePerTxn=0
    compaction_test.go:229: 
        	Error Trace:	pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go:229
        	Error:      	Not equal: 
        	            	expected: 25
        	            	actual  : 17
        	Test:       	TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=205/rowsDeletePerTxn=0
        	Messages:   	expected 25 number of wide scans issued, but 17 number of wide scan issued
    --- FAIL: TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=205/rowsDeletePerTxn=0 (16.68s)
=== RUN   TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1024
    compaction_test.go:229: 
        	Error Trace:	pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go:229
        	Error:      	Not equal: 
        	            	expected: 32
        	            	actual  : 24
        	Test:       	TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1024
        	Messages:   	expected 32 number of wide scans issued, but 24 number of wide scan issued
    --- FAIL: TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1024 (18.64s)
=== RUN   TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=2
    compaction_test.go:229: 
        	Error Trace:	pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go:229
        	Error:      	Not equal: 
        	            	expected: 32
        	            	actual  : 24
        	Test:       	TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=2
        	Messages:   	expected 32 number of wide scans issued, but 24 number of wide scan issued
    --- FAIL: TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=2 (32.83s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101767 pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test: TestSQLStatsCompactor failed [C-test-failure O-robot T-cluster-observability branch-master]
- #101674 pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test: TestSQLStatsCompactor failed [C-test-failure O-robot T-cluster-observability branch-release-23.1.0]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSQLStatsCompactor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test.TestSQLStatsCompactor [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9875633?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9875633?buildTab=artifacts#/) on release-23.1 @ [6944c293f14d5c1c5b28d6ea53329c3e214dd9ea](https://github.com/cockroachdb/cockroach/commits/6944c293f14d5c1c5b28d6ea53329c3e214dd9ea):


```
=== RUN   TestSQLStatsCompactor
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/9a30d5271a65be6b3f2aa1433be9011f/logTestSQLStatsCompactor1559819478
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestSQLStatsCompactor
    compaction_test.go:262: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9a30d5271a65be6b3f2aa1433be9011f/logTestSQLStatsCompactor1559819478
--- FAIL: TestSQLStatsCompactor (330.99s)
=== RUN   TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1
    compaction_test.go:229: 
        	Error Trace:	pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go:229
        	Error:      	Not equal: 
        	            	expected: 32
        	            	actual  : 27
        	Test:       	TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1
        	Messages:   	expected 32 number of wide scans issued, but 27 number of wide scan issued
    --- FAIL: TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1 (129.42s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101767 pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test: TestSQLStatsCompactor failed [C-test-failure O-robot T-cluster-observability branch-master]
- #101674 pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test: TestSQLStatsCompactor failed [C-test-failure O-robot T-cluster-observability branch-release-23.1.0]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSQLStatsCompactor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test.TestSQLStatsCompactor [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9898945?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9898945?buildTab=artifacts#/) on release-23.1 @ [82fddc52747a86c696a1f77a85abc25d05858cc1](https://github.com/cockroachdb/cockroach/commits/82fddc52747a86c696a1f77a85abc25d05858cc1):


```
=== RUN   TestSQLStatsCompactor
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/9a30d5271a65be6b3f2aa1433be9011f/logTestSQLStatsCompactor3142907628
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestSQLStatsCompactor
    compaction_test.go:262: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9a30d5271a65be6b3f2aa1433be9011f/logTestSQLStatsCompactor3142907628
--- FAIL: TestSQLStatsCompactor (334.67s)
=== RUN   TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1
    compaction_test.go:229: 
        	Error Trace:	pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test_test/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go:229
        	Error:      	Not equal: 
        	            	expected: 32
        	            	actual  : 25
        	Test:       	TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1
        	Messages:   	expected 32 number of wide scans issued, but 25 number of wide scan issued
    --- FAIL: TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=1 (130.31s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101767 pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test: TestSQLStatsCompactor failed [C-test-failure O-robot T-cluster-observability branch-master]
- #101674 pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test: TestSQLStatsCompactor failed [C-test-failure O-robot T-cluster-observability branch-release-23.1.0]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSQLStatsCompactor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/serverccl/statusccl/tenant_status_test.go
+++ b/pkg/ccl/serverccl/statusccl/tenant_status_test.go
@@ -132,6 +132,7 @@ func TestTenantStatusAPI(t *testing.T) {
 	})
 
 	t.Run("tenant_span_stats", func(t *testing.T) {
+		skip.UnderDeadlockWithIssue(t, 99770)
 		testTenantSpanStats(ctx, t, testHelper)
 	})
 

--- a/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go
+++ b/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go
@@ -143,6 +143,8 @@ func TestSQLStatsCompactor(t *testing.T) {
 
 	// Disable automatic flush since the test will handle the flush manually.
 	sqlConn.Exec(t, "SET CLUSTER SETTING sql.stats.flush.interval = '24h'")
+	// Disable activity update flush which also does a scan on the stats table
+	sqlConn.Exec(t, "SET CLUSTER SETTING sql.stats.activity.flush.enabled = false")
 	// Change the automatic compaction job to avoid it running during the test.
 	// Test creates a new compactor and calls it directly.
 	sqlConn.Exec(t, "SET CLUSTER SETTING sql.stats.cleanup.recurrence = '@yearly';")
@@ -221,9 +223,9 @@ func TestSQLStatsCompactor(t *testing.T) {
 
 			err = statsCompactor.DeleteOldestEntries(ctx)
 			kvInterceptor.disable()
+			expectedNumberOfWideScans := cleanupInterceptor.getExpectedNumberOfWideScans()
 			require.NoError(t, err)
 
-			expectedNumberOfWideScans := cleanupInterceptor.getExpectedNumberOfWideScans()
 			actualNumberOfWideScans := kvInterceptor.getTotalWideScans()
 
 			require.Equal(t,

